### PR TITLE
Refine Logic for Handling Nested Types

### PR DESCRIPTION
**Pull request by zhihuizhang17**
When handling nested types, there's no need to generate new symbol-kvs or file-kvs in the `state` for these nested types, as all their information is already stored in the root parent symbol's descriptor. Therefore:
+ For a nested type's key in state["symbols"] and state["files"], set the value to the parent's descriptor instead. 
+ Nested type dependencies are removed from the parent type.

By the way, this PR fixes issue #34.

### CODE REVIEW

File: priv/protos/test_service_v3.proto [hunk 1]
```diff
@@ -29,6 +29,11 @@ message TestRequest {
     }
 
     Location location = 2;
+    Token token = 3;
+  }
+
+  message Token {
```
Comment (mjheilmann) [inline priv/protos/test_service_v3.proto:35]: A nested payload is a message declaration inside of a message declaration, like
```
message A {
  B this = 0;
  message B {
    string message = 0;
  }
}
```

I'm not sure this effectively tests the parsing of nested messages

Comment (zhihuizhang17) [reply priv/protos/test_service_v3.proto:35]: `TestRequest.Payload` covers nested types tests.
`TestRequest.Token` covers following patterns.
```
message A {
  B b = 1;
  message B {
    C c = 1;
  }
  message C {
  }
}
```
In this pattern, Type-B and Type-C are both nested types in type-A, but B refers C.

Comment (mjheilmann) [reply priv/protos/test_service_v3.proto:35]: yup, `get_nested_types/2` is doing what I expected 👍

**Review by mjheilmann — APPROVED** [review]
If we aren't generating the symbol->descriptor map for nested types, doesn't that mean that we cannot `describe .namespace.ParentType.NestedType`?

If you `describe .namespace.ParentType` it includes the description of the nested type, so I'm not sure if this would be an actual issue

**Comment by zhihuizhang17** [comment]
> doesn't that mean that we cannot describe .namespace.ParentType.NestedType?

Yes, I think there's no need to do so. For a request like `{:file_containing_symbol, "a.b.C.D"}`, returning the descriptor file for "a.b.C" is sufficient. This solution aligns with Go's gRPC reflection. If you have concerns about this in Elixir, I can test it further with one of my online services later.

**Comment by mjheilmann** [comment]
> > doesn't that mean that we cannot describe .namespace.ParentType.NestedType?
> 
> Yes, I think there's no need to do so for all nested types. For a describe request like `{:file_containing_symbol, ".namespace.ParentType.NestedType"}`, returning the descriptor file for ".namespace.ParentType" is sufficient. This solution aligns with Go's gRPC reflection and it works with `grpcurl`. If you have concerns about this in Elixir, I can test it further with one of my online services later.

I agree that returning the payload from the parent type for the nested type is appropriate.  I mean that if we call `describe` on the nested symbol we will be unable to do so because we are not inserting the nested symbols into the State structure for later lookup in your changes.  You have code to try to do a dynamic lookup when searching the symbol like we do in the build stage, but I think keeping that only in the build stage is better.

**Review by mjheilmann — COMMENTED** [review]
Some of the concerns aren't as separated as they currently are on master:
- transform descriptors into our lookup structure in build
- reference the lookup structure on grpc call

Please also try to include tests for your changes.  You have test for your changes to `builder/util.ex` but you also modified `state.ex`

**Comment by zhihuizhang17** [comment]
> > > doesn't that mean that we cannot describe .namespace.ParentType.NestedType?
> > 
> > 
> > Yes, I think there's no need to do so for all nested types. For a describe request like `{:file_containing_symbol, ".namespace.ParentType.NestedType"}`, returning the descriptor file for ".namespace.ParentType" is sufficient. This solution aligns with Go's gRPC reflection and it works with `grpcurl`. If you have concerns about this in Elixir, I can test it further with one of my online services later.
> 
> I agree that returning the payload from the parent type for the nested type is appropriate. I mean that if we call `describe` on the nested symbol we will be unable to do so because we are not inserting the nested symbols into the State structure for later lookup in your changes. You have code to try to do a dynamic lookup when searching the symbol like we do in the build stage, but I think keeping that only in the build stage is better.

I understand your point. Let me rephrase my explanation:

It would be redundant to set a new `%Google.Protobuf.DescriptorProto` key-value pair for nested types in the generated Elixir code because the `%Google.Protobuf.DescriptorProto` struct for nested types is already contained within the parent's `%Google.Protobuf.DescriptorProto` struct.

Additionally, IMO, the `protoc-gen-elixir` plugin's approach of generating a `descriptor/0` function for all types, including nested types, is unnecessary and results in larger file sizes. This is indeed different from the approach taken in Go, where a single `fileDescriptor0` variable is generated for all types defined in the same `.proto` file.

I think there is an area for potential improvement in the `protoc-gen-elixir` plugin. It would be more efficient to generate the `descriptor/0` function only for top-level message types and enums, and rely on the nested message and enum definitions within the parent's `%Google.Protobuf.DescriptorProto` struct. This would not only reduce the generated code size but also align more closely with the approach taken by other language implementations of Protocol Buffers.

File: lib/grpc_reflection/service/state.ex [hunk 2]
```diff
@@ -56,8 +58,10 @@ defmodule GrpcReflection.Service.State do
   def lookup_symbol("." <> symbol, state), do: lookup_symbol(symbol, state)
 
   def lookup_symbol(symbol, %__MODULE__{symbols: symbols}) do
-    if Map.has_key?(symbols, symbol) do
-      {:ok, symbols[symbol]}
+    {_, root_symbol} = Util.get_package_and_root_symbol(symbol)
```
[the remark this answers is not in the document]
Comment (zhihuizhang17) [reply lib/grpc_reflection/service/state.ex:61]: > Anything that could be discovered here should instead be discovered during build and explicitly stored in the State object.

I got your point. It would be better to add a new nested map to the `state` struct, like this.
`%GrpcReflection.Service.State{root_symbols: %{}}`

File: lib/grpc_reflection/service/builder/util.ex [hunk 3]
```diff
@@ -7,20 +7,45 @@ defmodule GrpcReflection.Service.Builder.Util do
...
+
+  def get_parent_symbol(symbol) do
+    if String.contains?(symbol, ".") do
+      symbol |> String.split(".") |> Enum.slice(0..-2//1) |> Enum.join(".")
+    else
+      symbol
+    end
+  end
+
+  def get_package_and_root_symbol(symbol) do
```
[the remark this answers is not in the document]
Comment (zhihuizhang17) [reply lib/grpc_reflection/service/builder/util.ex:34]: > You have test for your changes to builder/util.ex but you also modified state.ex
Sure, I will add some targeted unit tests.

**Comment by mjheilmann** [comment]
@zhihuizhang17 Lets pause and rethink what is going on here

I think your core idea is correct, which means that I think there are 2 things and only 2 things that need to happen:
1. We don't list nested types as dependencies to be resolved later
1. We still want to be able to reflect nested types, we just return the parent payload instead of generating a payload for the nested type

Those 2 things, plus some tests, are all that is needed to correct the nested type problem.

> It would be redundant to set a new %Google.Protobuf.DescriptorProto key-value pair for nested types in the generated Elixir code because the %Google.Protobuf.DescriptorProto struct for nested types is already contained within the parent's %Google.Protobuf.DescriptorProto struct.

You've said this and things like this repeatedly in this PR.  You are missing the point of the State lookup object.  State already contains redundant information to streamline reflection calls.  State is comprised of several 1-1 lookup maps that align with the grpc reflection methods: `get_by_symbol` and `get_by_filename`.  [Reference again here](https://github.com/elixir-grpc/grpc-reflection/blob/main/lib/grpc_reflection/service/state.ex#L11).  There are a few other structures that assisted in building, such as `references` that could be removed and only present in a temporary structure used only in `build` but that is a different issue.

The state struct has the following traits that should be maintained:
1. `Map.keys(state.symbols)` lists every symbol that is in the given reflection tree
1. `Map.keys(state.files)` lists every pseudo proto file that we allow to be referenced in this reflection tree
1. The state object that is output from `build` is complete and static and fully resolves any reflection request -- or rejects as the case may be.

Optimizing denormalized payloads from State is out of the scope of correcting the nested types resolution problem.  I am open to this as a separate action, and have put some thought to this myself, but it needs to be done carefully as an independent PR.

Now as far as the dynamic lookup you added to `State.lookup_symbol`, this is problematic as it is a potential attack vector.  We perform the dynamic lookup in `build` because we have to due to the dynamic constraints of how the modules are associated and the `load_extensions` grpc problem.  `Code.ensure_loaded` can and will load and compile additional modules from disk, and this can have side effects on the runtime.  We are forced to do it during `build`, we should not do it anywhere else.

**Comment by zhihuizhang17** [comment]
> Just reusing the parent payload instead of generating new ones for nested types (current behavior) is already going to be a memory savings.

The current behavior generates a new `descriptor` (instead of reusing the parent type's descriptor) for nested types. This behavior caused the package name issue in your online service. I propose two solutions:

1. Don't store key-value pairs for nested types and return the parent's descriptor when a nested type's descriptor is requested.
2. For all nested types' keys, set the value to the parent's descriptor.

What do you think about these two solutions?

**Comment by mjheilmann** [comment]
Your proposal of option 2 is closest to what I've been asking for.

**Comment by zhihuizhang17** [comment]
> Your proposal of option 2 is closest to what I've been asking for.

Okay, I will address it in this PR this way.

File: lib/grpc_reflection/service/builder/util.ex [hunk 4]
```diff
@@ -7,6 +7,22 @@ defmodule GrpcReflection.Service.Builder.Util do
...
   @type_message Map.fetch!(Google.Protobuf.FieldDescriptorProto.Type.mapping(), :TYPE_MESSAGE)
 
+  def get_nested_types(symbol, descriptor), do: get_nested_types(symbol, descriptor, [])
+
+  def get_nested_types(_symbol, %Google.Protobuf.DescriptorProto{nested_type: []}, acc) do
+    acc
+  end
+
+  def get_nested_types(symbol, %Google.Protobuf.DescriptorProto{nested_type: nested_types}, acc) do
+    Enum.reduce(nested_types, acc, fn nested_type, acc ->
```
Comment (mjheilmann) [inline lib/grpc_reflection/service/builder/util.ex:17]: nit: you are recursing _and_ reducing here?  It seems to work, but usually you only see one at a time as they are different approaches to the same problem

**Review by mjheilmann — APPROVED** [review]

**Comment by mjheilmann** [comment]
Thanks for the work @zhihuizhang17
